packages/iris-grid/src/mousehandlers/IrisGridContextMenuHandler.tsx

```diff
@@ -919,7 +919,8 @@ class IrisGridContextMenuHandler extends GridMouseHandler {
       if (
         isEditableGridModel(model) &&
         model.isEditable &&
-        selectedRanges.length > 0
+        selectedRanges.length > 0 &&
+        model.isDeletableRanges(selectedRanges)
```

Comment:
From [Don's comment](https://github.com/deephaven/web-client-ui/pull/1996#issuecomment-2102734094):
> If we know it can never be shown for a given table it should be hidden rather than disabled. Only disable it if there's a condition in which it could be come enabled.

In this case you're just hiding it if the selected range isn't deletable, which if you have a totals row visible isn't quite correct - in that case, we should be showing the menu item, just disabled (since you _can_ delete some rows, just not the totals rows). Other tables where you can't delete anything is where it shouldn't be shown at all.

So there's a few different types of tables we should be testing here ([input_table docs here](https://deephaven.io/core/docs/how-to-guides/input-tables/)):
```python
from deephaven import agg, empty_table, input_table

# Immutable table - should not show Delete Selected Rows, should not be editable
immutable_table = empty_table(10).update(["X = i", "Y=Math.sin(i)"])

# Append only table - should be able to add new rows, Delete Selected Rows should not be visible
append_only_table = input_table(init_table=source)

# Keyed row table - should be able to add new rows, Delete Selected Rows should be visible
keyed_table = input_table(init_table=source, key_cols=["X"])
```

Then, after making the `keyed_table`, you should be able to go to the table menu, hit Aggregate Columns, and add an aggregation (such as Sum). (BUG NOTE: Looks like you'll need to position it at the "Top" or you can't see the totals - should log that as a separate issue). Right clicking a keyed row should show "Delete Selected Rows" and enabled, but right-clicking the totals row should just show "Delete Selected Rows" and have it disabled.

I think to achieve all that we'll change up the design a little bit... it will be a bit more complicated, but it will be a cleaner design.
- Add a `DeletableGridModel.ts` next to `EditableGridModel.ts`, and define an interface for a grid model that is deletable (e.g. an `isDeletable: boolean;` property, and `isDeletableRange` and `isDeletableRanges` methods). Also create the type predicates (`isDeletableGridModel` and `assertIsDeletableGridModel`. See `isEditableGridModel` and `assertIsEditableGridModel` in `EditableGridModel` for examples).
- Have `IrisGridProxyModel` implement `DeletableGridModel`, and delegate to it's underlying model if the underlying model is `DeletableGridModel` (much like the `EditableGridModel` methods work in `IrisGridProxyModel`)
- Have `IrisGridTableModelTemplate` implement the `DeletableGridModel`.
  - No need to change the other models, since they won't implement `DeletableGridModel`.
- Check `isDeleteable` for adding the menu, then disable with `!isDeletableRange`

Let me know if you have questions. Making this a little more complicated, but now it will be more robust.

Reply:
Also update your PR description after updating the content.